### ISSUE TITLE
Polyline info window bubble can drift when zooming

### ISSUE THREAD

**Issue by spyhunter99** [osmdroid/osmdroid#909]
## Issue Type

[x ] Bug

## Description and/or steps/code to reproduce the problem

See #865

**Comment by StevenTB** [comment]
I can confirm that updating from 5.6.5 to 6.0.2 breaks bubble attached to a marker when zooming or moving in the map

**Comment by b3nn0** [comment]
Hi,
not sure if it is the same issue, but I think it is. I just upgraded to OsmDroid 6 and noticed the same with a custom InfoWindow-like Class that I'm using. It uses MapView.LayoutParams.
The problem was introduced here I think:
https://github.com/osmdroid/osmdroid/commit/c6b48e6bb86c46eb0cc57563a213cdf7b21ddc00#diff-0bb50b730ea24896c5f4201928423da1
Several calls to requestLayout have been removed here, which I think is not correct.

Adding some code like this fixes the issue for me, but I think the requestLayout calls should come back:
```
        mapViw.addMapListener(new MapListener() {
            @Override
            public boolean onScroll(ScrollEvent event) {
                requestLayout();
                return false;
            }

            @Override
            public boolean onZoom(ZoomEvent event) {
                requestLayout();
                return false;
            }
        });
```

**Comment by monsieurtanuki** [comment]
@b3nn0 Hey, I did that!
From what I read in [SO](https://stackoverflow.com/questions/13856180/usage-of-forcelayout-requestlayout-and-invalidate), `requestLayout` is to be used when the `View`'s dimensions change, which is not the case in osmdroid when you zoom. And I removed those calls for performance reasons.

Btw I don't experience drifting info bubble on polyline ("More Samples / Drawing / Draw a polyline on screen") or marker ("More Samples / Data / Marker") when zooming. In which demo do you see that issue?

**Comment by b3nn0** [comment]
Hi,
well, you are right, usually requestLayout will only be needed when dimensions change. The call will cause onLayout to be called, which tells the view to layout its children.
In the case of a MapView, however, things might be different, as the child views should not stay at the same position relative to the view, but relative to the location on the map. I.e. when scrolling/zooming, the child views need to be repositioned on the screen, so that they scroll with the map.
As I said, I only observed the problem with my own InfoWindow-like class, which uses MapView.LayoutParams to position itself in the world. So I can only say, that MapView.LayoutParams behaves different to previous versions.
And as InfoWindow also uses MapView.LayoutParams, I simply assumed that it is the same problem.

However, looking at the code:
https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/views/overlay/infowindow/InfoWindow.java
I just noticed, that the LayoutParams are simply updated on each draw, which I guess fixes the problem aswell.

Anyway, I'm not sure how all this is supposed to be. In previous versions, requestLayout automatically made sure that MapView.LayoutParams only needs to be set once. In 6.*, this has changed and now I'm supposed to update the LayoutParams manually in each draw?
Not sure how the original Google Maps API handled this. I think the old behavior is more intuitive..

**Comment by monsieurtanuki** [comment]
@b3nn0 The fact is that performance is a huge factor in user experience, especially on smartphones. If we can get rid of unnecessary operations that are processed in _every_ draw, that's a good idea to remove them.

I understand that a side-effect of this unnecessary `requestLayout` call was that - in your case - the drawing parameters of your own info window were correctly re-computed.

I'm glad you found the perfect solution to make your app work correctly: refresh the info window layout param offset when you draw your info window. This way you can enjoy better performances when you don't have an info window, and an appropriate info window positioning.

@spyhunter99 @StevenTB Could you please tell me where in the demo app you see that strange behavior, for marker or for polylines?

**Comment by spyhunter99** [comment]
Drawing, draw a polyline on screen.

Draw the line, then keep zooming in.

i think the issue is that the anchor point of the info window is based on where the user tapped the screen and not anchored to an actual point on the line. I tried to implement something to find the closest point point on the line to the tap location but i wasn't able to get the math right

**Comment by monsieurtanuki** [comment]
Fixed by #1136

**Comment by monsieurtanuki** [comment]
@spyhunter99 Ready to test if #1136 fixed this issue?

**Comment by monsieurtanuki** [comment]
I've just merged #1195, that puts back the calls to method `mapView.requestLayout`, which should solve the initial issue and which anyway restores the way it worked at 5.6.5.
@spyhunter99  @StevenTB  @b3nn0 Feel free to test.
